[4-9단계 - 방탈출 예약 관리] 듀이(이주희) 미션 제출합니다.

ljhee92 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  https://prolog.techcourse.co.kr/studylogs/4162

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- 책임에 따라 클래스를 분리하려고 노력했습니다.
- 모든 원시값을 포장하지는 않았습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [x] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [ ] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
범블비 안녕하세요!
이번 리뷰도 잘 부탁드립니다😄

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

ljhee92 committed:
refactor: 테스트 코드 테스트명 구체화

ljhee92 committed:
refactor: delete 요청 시 id가 없다면 noContent(204) 상태를 명시하여 반환하도록 수정

ljhee92 committed:
refactor: String이 아닌 DateTimeFormatter 객체 자체를 상수로 처리

ljhee92 committed:
feat: 이름, 예약일자, 예약시간 유효성 검증 기능 추가

ljhee92 committed:
feat: 데이터베이스 테이블 스키마 정의

ljhee92 committed:
feat: 데이터베이스 설정

ljhee92 committed:
test: 데이터베이스 접근, 데이터베이스명, 테이블명 확인 테스트

ljhee92 committed:
chore: date, time 컬럼 datetime으로 통합하여 관리

- date, time을 따로 저장하는 것보다 datetime으로 저장하는 것이 데이터의 일관성을 더 보장할 수 있다고 판단하여 변경

ljhee92 committed:
feat: DB 적용 후 데이터 조회 기능 추가

- Repository는 도메인 관점 저장소, Dao는 h2와 연결되는 데이터 연결 객체이므로 패키지 분리

ljhee92 committed:
chore: 요구사항에 따라 기존에 사용하던 repository 제거

ljhee92 committed:
feat: DB 적용 후 데이터 추가, 삭제 기능 추가

ljhee92 committed:
refactor: RowMapper로 중복제거

ljhee92 committed:
refactor: 데이터 추가시 KeyHolder -> SimpleJdbcInsert로 수정

ljhee92 committed:
refactor: ReservationDao 단위테스트를 위해 interface로 추상화, Reservation 정적 팩토리 메서드로 생성하도록 변경

- ReservationDao를 interface로 추상화하고 Fake 객체로 구현하여 단위 테스트 진행
- db에서 id를 자동생성하기 전에 id가 없는 Reservation 생성해야 하는데, 외부에서 null을 주입하여 생성해주는 것보다 Reservation 내부에서 정적 팩토리 메서드로 생성하는 것이 통일성 있다고 판단하여 변경

ljhee92 committed:
refactor: 매개변수 final 통일

ljhee92 committed:
test: 테스트용 데이터베이스 분리 및 단위 테스트 구현

ljhee92 committed:
docs: 시간 관리 기능 명세 추가

ljhee92 committed:
feat: 예약 시간 유효성 검증 기능 추가

ljhee92 committed:
feat: 시간 조회 기능 추가

ljhee92 committed:
feat: 시간 추가 기능 추가

ljhee92 committed:
feat: 시간 삭제 기능 추가

ljhee92 committed:
docs: 시간 선택 기능 명세 추가

ljhee92 committed:
refactor: 요구사항에 맞춰 스키마의 datetime -> date, time으로 분리, 도메인의 LocalDateTime -> LocalDate, LocalTime으로 분리

ljhee92 committed:
feat: 예약 페이지 매핑 수정

ljhee92 committed:
feat: 예약시간 값 객체 생성

ljhee92 committed:
feat: 특정 예약시간 조회 기능 추가

ljhee92 committed:
chore: 테스트 시 사용하는 스키마, 데이터 파일 명시, 예약 시간 데이터 추가

ljhee92 committed:
feat: 예약시 예약시간을 지정된 값에서 선택하는 기능 추가

ljhee92 committed:
fix: 조회 시 Response API 명세에 맞추어 수정

ljhee92 committed:
test: controller와 dao 책임 분리 테스트 추가

ljhee92 committed:
refactor: 메서드명 명확하게 변경

ljhee92 committed:
refactor: rowMapper로 중복 매핑 제거

ljhee92 committed:
refactor: 내장 임베디드 데이터베이스가 아닌, 설정해둔 testDatabase를 사용하도록 설정

ljhee92 commented on src/main/java/roomescape/controller/ReservationController.java:
```diff
@@ -36,17 +42,16 @@ public ResponseEntity<List<ReservationResponse>> getReservations(){
-        Reservation reservation = reservationRepository.add(reservationRequest);
-        return ResponseEntity.ok().body(ReservationResponse.toDto(reservation));
+        ReservationTime reservationTime = reservationTimeDao.findById(reservationRequest.timeId());
+        Reservation newReservation = reservationDao.insert(reservationRequest.toEntity(reservationTime));
```
`Reservation`을 생성할 때 `ReservationTime`을 조회하고, 주입해주는 과정이 `Controller`의 책임이 맞을까? 라는 생각 때문에 아래 코드처럼 `Service`를 도입해 분리하려고 했었습니다.
```java
public ReservationResponse createReservation(final ReservationRequest reservationRequest) {
        ReservationTime reservationTime = reservationTimeDao.findById(reservationRequest.timeId());
        Reservation reservation = Reservation.of(reservationRequest.name(), reservationRequest.date(), reservationTime);
        Reservation newReservation = reservationDao.insert(reservation);
        ReservationTimeResponse reservationTimeResponse = new ReservationTimeResponse(newReservation.getReservationTime().getId(), newReservation.getReservationTime().getStartAt().format(TIME_FORMATTER));
        String reservationDateResponse = newReservation.getReservationDate().format(DATE_FORMATTER);
        return new ReservationResponse(newReservation.getId(), newReservation.getCustomerName(), reservationDateResponse, reservationTimeResponse);
}
```
그런데 Dto <-> 도메인엔티티 간 변환 과정, 조회한 ReservationTime을 통해 Reservation을 조립하는 과정이 모두 Service의 책임인 비즈니스 로직이 맞는지 잘 모르겠습니다.🥲

ddaaac replied:
거꾸로 질문을 해볼 수 있을 것 같아요.

> Dto <-> 도메인엔티티 간 변환 과정, 조회한 ReservationTime을 통해 Reservation을 조립

이건 어떤 객체나 레이어에서는 꼭 해야되는 과정같아보여요. 지금은 어디서 이 행위를 하고 있을까요?
듀이가 생각하시기에 적절한 책임이라고 생각하나요?

ljhee92 replied:
기존에 Dto <-> 도메인엔티티 간 변환 과정은 DTO 내부에서,
조회한 `ReservationTime`을 통해 `Reservation`을 조립하는 것은 `Controller`에서 수행하고 있었습니다.

> Reservation을 생성할 때 ReservationTime을 조회하고, 주입해주는 과정이 Controller의 책임이 맞을까?

라는 생각 때문에 `Service`를 도입하고자 했으면서 Dto <-> 도메인엔티티 변환까지 `Service`에 넣으려고 해서 거부감이 들었던 것 같아요.
Dto와 도메인엔티티 변환을 `Service`에서 수행하게 된다면 파라미터가 늘어날 때 서비스는 더 책임이 많아질 것 같고, 현재로서는 DTO 내부에 두는 것이 더 명확한 책임 분리라고 생각합니다.
그래서 기존에 컨트롤러에서 분리하고자 했던 부분만 `Service`로 분리하였고, `ReservationTimeController` 에서는 `ReservationController`처럼 별도의 분리해야 할 비즈니스 로직이 있다고 생각하지는 않지만 일관성을 위해 계층을 함께 분리해주었습니다.

[refactor: Service 계층 도입](https://github.com/ljhee92/spring-roomescape-admin/commit/77344c4b23edae32fb5854bad1406d2e185b0516)

ljhee92 committed:
refactor: 마지막 개행 추가

ljhee92 committed:
fix: Test 클래스마다 db 격리를 위해 @ActiveProfiles, @AutoConfigureTestDatabase 어노테이션 제거

ljhee92 commented:
테스트 클래스마다 `application-test.properties`로 설정해둔 testDatabase를 사용하고 싶어서 어노테이션을 붙여두었는데, 그렇게 되니 클래스끼리 데이터베이스가 공유되어 독립적이지 못해 테스트가 깨지는 현상이 생겼습니다.
테스트가 깨지는지 모르고 있다가 리뷰요청 후에 발견해서 급하게 추가로 푸시하나 했습니다!🥹

ljhee92 committed:
refactor: 생성자가 하나일 경우 @Autowired 생략 가능

ljhee92 commented:
테스트 클래스마다 application-test.properties로 설정해둔 testDatabase를 사용하고 싶어서 어노테이션을 붙여두었는데, 그렇게 되니 클래스끼리 데이터베이스가 공유되어 독립적이지 못해 테스트가 깨지는 현상이 생겼습니다.
테스트가 깨지는지 모르고 있다가 리뷰요청 후에 발견해서 급하게 추가로 푸시 했습니다!🥹

ddaaac commented on src/main/java/roomescape/entity/Reservation.java:
```diff
@@ -1,28 +1,60 @@
+        this.reservationTime = reservationTime;
+    }
+
+    public static Reservation of(final String customerName, final LocalDate reservationDate, final ReservationTime reservationTime) {
+        return new Reservation(null, customerName, reservationDate, reservationTime);
+    }
+
+    public static Reservation of(final Long id, final String customerName, final LocalDate reservationDate, final ReservationTime reservationTime) {
+        return new Reservation(id, customerName, reservationDate, reservationTime);
+    }
```
[생성자 체이닝](https://developer-talk.tistory.com/227)을 활용할 수 있을 것 같습니다.

ljhee92 replied:
정적 팩토리 메서드는 기존에 알고 있어서 활용해보았는데, 생성자 체이닝은 몰랐던 부분입니다.
범블비가 알려주신 링크도 읽어 보았고, 제가 별도로 [생성자 체이닝](https://www.baeldung.com/java-chain-constructors)에 대해서도 찾아보기도 했는데요.
- `null` 처리를 생성자 내부에서 할 수 있다.
- 객체 생성 시 `this`를 통해 하나의 생성자에서만 필드 초기화가 가능하므로 중복을 제거할 수 있다.
는 장점이 있다고 생각하여 적용해보았습니다👍🏻

[refactor: 생성자 체이닝 적용](https://github.com/ljhee92/spring-roomescape-admin/commit/e0b763777ecaad58832bb5a2abef365a176dd0d0)

ddaaac commented on src/main/java/roomescape/entity/Reservation.java:
```diff
@@ -1,28 +1,60 @@
+    }
+
+    private void validateReservationDate(final LocalDate reservationDate) {
+        if (reservationDate == null) {
```
보통 자바에서 필드나 파라미터가 non-nullable하다고 가정합니다. 오히려 반대로 nullable한 경우를 신경써주는 게 맞겠죠.

ljhee92 replied:
범블비의 피드백을 받고 생각해보니 `null` 검증을 도메인에서 할 것이 아니라 DTO를 생성할 때 하는 것이 더 적절하다고 생각했습니다.
그래서 `Reservation`, `ReservationTime` 내부에 있던 `null` 검증을 `ReservationRequest`, `ReservationTimeRequest`가 생성될 때 검증하도록 이동시켰습니다.

검증 로직을 이동시키면서 발생한 예외를 처리해주지 않으면 스프링에서는 클라이언트에게 500(내부 서버 오류) 코드를 반환한다는 것을 알게 되었고, 해당 예외는 500보다 400(클라이언트의 잘못된 요청) 코드가 더 알맞다고 생각했고, 이를 명시해주기 위해 `ExceptionHandler` 어노테이션을 활용해서 예외 처리를 해주었습니다. 

또, `ReservationRequest` 생성시 `name`, `date`의 `null` 검증은 해주고 있으면서, `timeId`의 `null` 검증을 하지 않았는데요. `timeId`를 클라이언트에서 선택해서 요청하기 때문에 `null` 값이 들어올 일이 없다고 판단했기 때문에 별도로 검증하지 않았습니다.

[refactor: null 검증 DTO로 이동, ExceptionHandler로 예외처리](https://github.com/ljhee92/spring-roomescape-admin/commit/42fa0d9d96345e43e2f7d4b6e03e1235b1fead37)

ddaaac commented on src/test/resources/schema-test.sql:
```diff
@@ -0,0 +1,16 @@
+CREATE TABLE IF NOT EXISTS reservation_time
```
test 프로퍼티를 분리해주셨군요 👍 

schema 같은 경우에는 test나 리얼 환경이나 항상 동일해야할 것 같아요.

ljhee92 replied:
단순히 test 프로퍼티를 분리해야 한다는 생각 때문에 같은 schema를 사용하고 있음에도 불구하고 별도로 처리를 해주고 있었네요!
schema 같은 경우에는 프로덕션 환경이나 테스트 환경에서 항상 동일해야 한다는 범블비 말씀에 동의합니다.

`application-test.properties`에서 테스트에 사용할 schema를 명시해주지 않으면 main/resources에 있는 schema를 사용한다고 알고 있어 test/resources에 있는 `schema-test.sql` 파일을 제거하고, 프로퍼티 내부에서도 스키마 경로를 명시하고 있던 부분을 삭제해주었습니다.

[chore: 프로덕션과 테스트 환경의 schema가 항상 동일하도록 처리](https://github.com/ljhee92/spring-roomescape-admin/commit/145c864c5bc61a0dcfa3d29107d31efd0cb8cfad)

ddaaac commented on src/main/java/roomescape/dto/ReservationRequest.java:
```diff
@@ -1,14 +1,13 @@
-import java.time.LocalTime;
 
-public record ReservationRequest(String name, LocalDate date, LocalTime time) {
+public record ReservationRequest(String name, LocalDate date, Long timeId) {
```
Request에서 LocalTime으로의 변환은 어떻게 되는걸까요? 클라이언트에서는 어떤 규격에 맞춰서 요청해야할까요?

ljhee92 replied:
클라이언트는 API 명세에 따라 `timeId`를 요청하고 있습니다.
서버에서는 `timeId`를 받아 `H2ReservationTimeDao#findById`에서 DB의 값을 조회하고, 조회한 값을 `RowMapper`로 `LocalTime` -> `ReservationTime`으로 변환하고 있습니다.

범블비가 어떤 의도로 질문을 주신 건지 잘 이해하지 못해서 코드에 어떻게 반영해야할 지 모르겠습니다..!
다시 한 번 설명해주실 수 있을까요~?

ddaaac replied:
아 뭔가 반영하라고 남긴 코멘트는 아니었고 조금 더 공부해봤으면 하는 포인트를 남겼었습니다.

[링크](https://ksh-coding.tistory.com/107)에서 학습해보시면 좋을 것 같아요.

ljhee92 replied:
`@ResponseBody`, `@RequestBody`를 사용하게 되면 내부적으로 `Jackson` 라이브러리를 사용해서 직렬화/역직렬화되기 때문에 기본적으로 지정된 패턴과 다른 패턴으로 고정해야 한다면 `@JsonFormat(pattern = "HH:mm")`과 같이 지정해주어야 명확하게 바인딩이 되겠네요!
현재 `LocalTime`의 경우 `HH:mm` 형식으로도 잘 바인딩 되고 있지만 명확하게 해주기 위해서 `@JsonFormat` 어노테이션으로 정확하게 명시해주었습니다.
추가로, `ReservationTimeResponse`에서도 `DateTimeFormatter` -> `String`으로 변환하여 반환하지 않고, `@JsonFormat`을 활용해서 `LocalTime` 타입으로 반환하도록 바인딩해주었고, 마찬가지로 `ReservationResponse`에서도 `LocalDate`를 반환할 수 있게 했습니다.
`LocalDate`는 기본 형식인 'yyyy-mm-dd'에서 바뀌지 않았기 때문에 `@JsonFormat`으로 형식을 명시해주지 않았습니다.

범블비 덕분에 `@JsonFormat` 사용법, 직렬화/역직렬화 과정에 대해서도 좀 더 공부할 수 있었습니다!

[refactor: JsonFormat annotation 사용하여 LocalTime 형식 명시, ReservationResponse에서 LocalDate 반환하도록 변경](https://github.com/ljhee92/spring-roomescape-admin/commit/88cd0008c92cc3d3b1cd60b88dc183bbaffe1670)

ddaaac reviewed (changes requested):
안녕하세요 듀이 👋 

간단하게 잘 구현해주셔서 크게 남길 코멘트가 없네요.

몇 가지 코멘트 확인해주세요.

ljhee92 committed:
refactor: 생성자 체이닝 적용

생성자 체이닝의 장점
- null 처리를 생성자 내부에서 할 수 있다.
- 객체 생성시 this를 통해 하나의 생성자에서만 필드 초기화가 가능하므로 중복을 제거할 수 있다.

ljhee92 committed:
refactor: null 검증 DTO로 이동, ExceptionHandler로 예외처리

- null 검증은 도메인이 생성될 때보다 DTO가 생성될 때 이루어지는 것이 더 적합하다고 판단되어 검증 로직을 이동했습니다.
- 예외를 처리해주지 않으면 500(서버 내부 오류)코드로 반환되기 때문에 400(클라이언트의 잘못된 요청)이라고 명시해주기 위해 ExceptionHandler로 처리해주었습니다.
- ReservationRequest 생성 시 timeId는 선택값으로 들어오기 때문에 null이 들어올 수 없다고 판단하여 검증하지 않았습니다.

ljhee92 committed:
chore: 프로덕션과 테스트 환경의 schema가 항상 동일하도록 처리

ljhee92 committed:
refactor: Service 계층 도입

- Reservation 생성시 timeId로 ReservationTime을 조회하고 생성하는 로직을 비즈니스 로직이라고 여겨 Controller에서 수행하지 않고 Service 계층을 도입해 수행했습니다.
- ReservationTimeController에서는 별도의 비즈니스 로직이 있다고 생각하지 않았지만, 일관성 있게 계층을 맞춰주기 위해 Service를 함께 도입했습니다.

ljhee92 committed:
refactor: fake 객체 패키지 분리

ljhee92 commented:
안녕하세요 범블비~!
주신 리뷰 반영해보았습니다.
혹시 제가 잘못 이해하고 반영한 부분이 있다면 말씀해주세요!
소중한 시간 들여 리뷰해주셔서 감사합니다!😄👍🏻

ddaaac commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,39 @@
+    public ReservationResponse createReservation(final ReservationRequest reservationRequest) {
+        ReservationTime reservationTime = reservationTimeDao.findById(reservationRequest.timeId());
+        Reservation newReservation = reservationDao.insert(reservationRequest.toEntity(reservationTime));
+        return ReservationResponse.toDto(newReservation);
```
A라는 예약자는 한 타임에 중복된 예약을 할 수 없다 라는 요구사항을 추가해봐도 좋을 것 같아요!

ljhee92 replied:
앗 넵!
범블비가 말씀해주신 A라는 동일한 예약자가 같은 날짜, 같은 시간에 중복된 예약을 할 수 없다는 요구사항 추가하고 구현해보았습니다.
해당 기능 구현 전에, 중복된 예약 시간을 추가할 수 없는 기능을 먼저 추가했습니다.
이 기능과 관련된 커밋이 나뉘어져 있어서 조금 많은데요!
(커밋 번호 3b82c6580d2a70c363212f9433d0b154ace37149, e3763942f270626cf1de797debc2935258b8bf0b, c1cc33aa052272c4a7a45122b5effededda56baf, a738cb0664054285debd872187a80469c08400ad, fba360ecf34f875317075acf3efa64808eea120c, 82b069228f6248120f3ec205ad81086ce65074a2)
최종으로 동작하는 코드를 확인하시려면 [fix: 테스트 예약 시간 고정](https://github.com/ljhee92/spring-roomescape-admin/commit/82b069228f6248120f3ec205ad81086ce65074a2) 본 커밋 시점에서 확인 부탁드립니다!

ddaaac commented on src/test/java/roomescape/service/ReservationServiceTest.java:
```diff
@@ -0,0 +1,49 @@
+
+    @BeforeEach
+    void setUp() {
+        ReservationDao reservationDao = new FakeReservationDao();
+        ReservationTimeDao reservationTimeDao = new FakeReservationTimeDao();
+        reservationService = new ReservationService(reservationDao, reservationTimeDao);
```
mock 객체와 fake 객체의 방식의 차이에 대해서도 알아보면 좋을 것 같습니다.

ljhee92 replied:
레벨1 장기 미션 때 `Fake` 객체를 처음 사용하면서 `Fake` 객체에 대해서는 확실히 이해했고, 이번 미션에서도 활용해볼 수 있었습니다.
이 때 테스트 더블에 대해 대략적으로만 알고 각각의 차이는 학습하지 않았었는데요.
이번 기회에 `Mock` 객체와 `Fake` 객체에 대해 학습해보았습니다.

`Fake` 객체는 현재 테스트 코드에서 구현한 `FakeReservationDao`, `FakeReservationTimeDao` 처럼 **진짜 DB와 연결되지는 않지만 연결된 것처럼 동작하는 객체**입니다.
`Mock` 객체는 `Mockito`라는 라이브러리를 사용해서 **호출 횟수, 호출 여부, 반환 값 등을 통제**할 수 있습니다. 실제처럼 동작하는 `Fake` 객체와 다르게 **껍데기만 있는 객체**입니다. 그렇기 때문에 `Fake` 객체와 달리 세부 구현에 영향을 받지 않고, 테스트가 실패해야 하는 경우에도 성공하는 등 오류가 발생할 가능성이 높습니다.
단 `Fake` 객체를 직접 구현해야 할 경우 복잡한 환경일 수록 시간이 오래 걸릴 수 있어서, `Mock` 객체를 활용하게 된다면 훨씬 빠르게 테스트가 가능합니다.

아직 `Mock` 객체를 활용해서 테스트 코드를 작성해본 적은 없어서 좋은 코드, 나쁜 코드 책에 있는 예시와 함께 몇몇 크루들과 이야기하여 학습해보았습니다!

ddaaac replied:
잘 정리해주셨습니다.

> Fake 객체와 달리 세부 구현에 영향을 받지 않고, 테스트가 실패해야 하는 경우에도 성공하는 등 오류가 발생할 가능성이 높습니다.

mock 객체는 객체 내부 구현과 굉장히 밀접하게 연관됩니다. 내부 구현이 바뀌면 테스트 코드도 같이 영향을 받구요. 다음 미션이나 언젠가 한 번 직접 사용해보시면 좋을 것 같습니다! 두 방식의 차이점을 생각해주시면 좋을 것 같아요.

ddaaac reviewed (changes requested):
안녕하세요 듀이 👋 

리뷰 반영 잘 해주셨습니다.

몇 가지 코멘트 추가로 남겼으니 확인해주세요.

ljhee92 committed:
refactor: JsonFormat annotation 사용하여 LocalTime 형식 명시, ReservationResponse에서 LocalDate 반환하도록 변경

ljhee92 committed:
docs: 요구사항 추가(시간 추가 시 중복 검증, 예약 추가 시 중복 검증)

ljhee92 committed:
feat: 예약 시간 중복 검증 기능 추가

ljhee92 committed:
refactor: 테스트케이스 분리

ljhee92 committed:
refactor: 메서드명, 매개변수 수정

ljhee92 committed:
feat: 동일한 예약자 예약날짜, 시간 중복 예약 불가 기능 추가

ljhee92 committed:
fix: 테스트 예약 시간 고정

ljhee92 commented:
안녕하세요 범블비!
주말에도 리뷰해주셔서 감사합니다👍🏻 피드백해주신 부분 반영했습니다.
제가 이전 리뷰요청 때 코멘트를 먼저 작성하고 커밋내역을 코멘트에 추가한다는게 깜빡했더라구요😭
이번 요청 때에는 편하게 확인하실 수 있도록 커밋내역 함께 추가해두었습니다!

ddaaac commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,43 @@
+    public ReservationResponse createReservation(final ReservationRequest reservationRequest) {
+        ReservationTime reservationTime = reservationTimeDao.findById(reservationRequest.timeId());
+        Reservation reservation = reservationRequest.toEntity(reservationTime);
+        if (reservationDao.duplicateReservationByCustomer(reservation, reservationTime)) {
+            throw new IllegalArgumentException("[ERROR] 같은 예약자가 같은 날짜, 시간에 중복으로 예약할 수 없습니다.");
+        }
```
중복 처리 잘 해주셨네요. DB의 유니크 인덱스를 활용해볼 수도 있을 거 같아요. 방식은 다양하겠죠.

여튼, 중복 처리를 해봤으면 좋겠다는 코멘트를 남긴 이유는 서비스 계층이 갖는 책임과 역할에 대해서 고민해보면 좋을 것 같아서였습니다.

지금이야 요구사항이 워낙 간단하니 서비스 계층의 필요성을 잘 못 느꼈을텐데요, 요구사항이 계속 추가되고 그 요구사항을 도메인만으로 풀기 어려운 경우가 발생합니다.
지금처럼 DB에 저장되어 있는 객체가 있을 때가 있겠죠. 이런 경우에 서비스 계층에서 인프라와 도메인을 연결해주는(객체지향적으로 봤을 때는 메세지를 전달해주는) 역할을 할 수 있겠죠.

앞으로는 더 많은 요구사항에 부딪힐텐데 그때그때 적당한 방식을 고민해보시면 좋을 것 같습니다.

ddaaac reviewed (approved):
안녕하세요 듀이 👋 

리뷰 반영 잘 해주셨습니다!

다음 미션도 화이팅하세요.
